openhands/agenthub/codeact_agent/codeact_agent.py

```diff
@@ -88,66 +93,165 @@ def __init__(
             if config.micro_agent_name
             else None
         )
+        if (
+            self.config.function_calling
+            and not self.llm.config.supports_function_calling
+        ):
+            logger.warning(
+                f'Function calling not supported for model {self.llm.config.model}. '
+                'Disabling function calling.'
+            )
+            self.config.function_calling = False
+
+        if self.config.function_calling:
```

Comment:
all the `if/else` function_calling logic here is a little ugly. I wonder if we just need a new CodeActFunctionAgent which inherits from CodeAct?

Not for this PR necessarily, but it seems like the logic diverges quite a bit

Reply:
Totally agree here! I think that makes sense - I can put up a follow-up PR to clean this up a bit.